### PR TITLE
Feature/api list applications endpoint

### PR DESCRIPTION
**Pull request by lorenzofk**
**What does this PR do?**

This PR implements a new API endpoint (Task 1) that allows authenticated users to retrieve a paginated list of applications. The users can optionally filter the applications by plan type (nbn, opticomm, or mobile).

**Key Changes:**

- Create the controller and the action to handle the API endpoint;
- Create a new FormRequest file to validate the request payload;
- Create a new API Resource to isolate the logic to build the response data;
- Create a new config file to store the pagination setting (number of items per page);
- Create the model relationships and accessors (monthly cost, full address and full name) in different models;
- Implement unit and feature tests to cover different areas of the system.

The API endpoint should now be paginated and only accessible for authenticated users.

**Additional Information:**

Screenshot 2025-02-06 at 10 50 55 pm

### CODE REVIEW

File: app/Http/Controllers/ApplicationController.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+class ApplicationController extends Controller
+{
+    /**
+     * Return a list of applications.
+     */
+    public function index(ListApplicationsRequest $request): AnonymousResourceCollection
+    {
+        $planType = $request->query('plan_type');
+
+        $applications = Application::when($planType, fn ($query) => $query->planType($planType))
```
Comment (lorenzofk) [inline app/Http/Controllers/ApplicationController.php:19]: I considered to move this logic to a Service layer. Although I think it might be a bit overkilling as this endpoint is pretty simple.

File: app/Http/Requests/ListApplicationsRequest.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+     *
+     * @return array<string, \Illuminate\Contracts\Validation\Rule|array|string>
+     */
+    public function rules(): array
+    {
+        return [
+            'plan_type' => [
+                'nullable',
+                'string',
+                Rule::in(['nbn', 'mobile', 'opticomm']),
```
Comment (lorenzofk) [inline app/Http/Requests/ListApplicationsRequest.php:29]: I considered to have an Enum to represent the Plan's type, but decided to leave it as it is for time constraints and because the provided factory was already using hard-coded values.

File: app/Models/Application.php [hunk 3]
```diff
@@ -18,4 +21,40 @@ class Application extends Model
...
+    {
+        return $this->belongsTo(Customer::class);
+    }
+
+    /**
+     * Scope a query to only include applications with a plan of a given type.
+     */
+    public function scopePlanType(Builder $query, string $planType): Builder
+    {
+        return $query->withWhereHas('plan', fn ($query) => $query->where('type', $planType));
```
Comment (lorenzofk) [inline app/Models/Application.php:40]: This method should ensure the relationship exists and also eager loads it.

File: tests/Unit/Models/ApplicationTest.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        // Filter by `mobile` plan type
+        $result = Application::planType('mobile')->get();
+
+        $this->assertCount(1, $result);
+        $this->assertEquals($mobileApplication->id, $result->first()->id);
+    }
+
+    public function test_should_return_the_full_address_of_the_application(): void
+    {
+        $application = Application::factory()->make([
```
Comment (lorenzofk) [inline tests/Unit/Models/ApplicationTest.php:45]: Intentionally used `make` here to avoid creating a new record just to test an accessor. The `make` still creates the internal relationships, which is a bit annoying. But I decided to kept it anyway to avoid creating a new instance of the model manually and keep using the same "standard" from other test cases.

File: tests/Feature/Application/ApiTest.php [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+    private User $user;
+
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $this->user = User::factory()->create();
+
+        $this->actingAs($this->user);
```
Comment (lorenzofk) [inline tests/Feature/Application/ApiTest.php:30]: As the requirements are saying that `This endpoint will exist only for authenticated users` I assume that the user will be always authenticated in this test suite. I could have added an edge case test to ensure that the endpoint returns `forbidden` when a guest hits it.

File: tests/Feature/Application/ApiTest.php [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+                )
+                 ->has('data.1', fn (AssertableJson $json) =>
+                    $json->where('application_id', $completeApplication->id)
+                        ->where('customer_name', $completeApplication->customer->full_name)
+                        ->where('address', $completeApplication->full_address)
+                        ->where('state', $completeApplication->state)
+                        ->where('plan_type', $completeApplication->plan->type)
+                        ->where('plan_name', $completeApplication->plan->name)
+                        ->where('plan_monthly_cost', $completeApplication->plan->monthly_cost_formatted)
+                        ->has('order_id')
```
Comment (lorenzofk) [inline tests/Feature/Application/ApiTest.php:108]: The main goal here of checking both of the applications from the response is to show that they have different fields (order_id) depending on the status.

File: tests/Feature/Application/ApiTest.php [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+    public function test_should_return_the_applications_sorted_by_oldest_first(): void
+    {
+        // Create two applications with different `created_at` dates
+        $applications = Application::factory()
+            ->count(2)
+            ->create([
+                'created_at' => new Sequence(now()->subDays(2), now())
+            ])
+            ->sortBy('created_at')
```
Comment (lorenzofk) [inline tests/Feature/Application/ApiTest.php:133]: When creating a factory, the `Sequence` guarantees the records are stored in that specific order. However, sometimes at the time to retrieve the data, it gets returned in a different order. So I decided to make sure it's always correctly sorted to prevent unexpected issues.

File: tests/Unit/FormRequest/ListApplicationsRequestTest.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    private ListApplicationsRequest $request;
+
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $this->request = new ListApplicationsRequest();
+    }
+
+    public function test_should_pass_without_any_plan_type(): void
```
Comment (lorenzofk) [inline tests/Unit/FormRequest/ListApplicationsRequestTest.php:23]: I decided to keep this test suite simple and avoided using data providers here, as each test has a potential different expected result, so it would simplify some parts of the class, but it would make the test itself a bit complex with conditionals and different flows.

File: tests/Unit/Resources/ApplicationResourceTest.php [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        $data = [
+            'application_id' => $application->id,
+            'customer_name' => $application->customer->full_name,
+            'address' => $application->full_address,
+            'state' => $application->state,
+            'plan_type' => $application->plan->type,
+            'plan_name' => $application->plan->name,
+            'plan_monthly_cost' => $application->plan->monthly_cost_formatted,
+            'order_id' => $application->status === ApplicationStatus::Complete ? $application->order_id : new MissingValue,
```
Comment (lorenzofk) [inline tests/Unit/Resources/ApplicationResourceTest.php:55]: The `MissingValue` here means that the value will be omitted from the request when it's parsed, which is exactly what we want to do for applications that are not `completed`.
